packages/engine/src/worker/runner/javascript/mod.rs

```diff
@@ -538,162 +539,118 @@ impl<'m> RunnerImpl<'m> {
+        field: &Field,
         len: Option<usize>,
     ) -> Result<ArrayData> {
         // `data` must not be dropped until flush is over, because
         // pointers returned from FFI point inside `data`'s ArrayBuffers' memory.
-        let obj = data
-            .as_object()
-            .ok_or_else(|| Error::Embedded("Flush data not object".into()))?;
-        let child_data: mv8::Array<'_> = obj.get("child_data")?;
+        let obj = data.as_object().ok_or_else(|| {
+            Error::Embedded(format!("Flush data not object for field {:?}", field))
+        })?;
 
         // `data_node_from_js` isn't recursive -- doesn't convert children.
         let data: mv8::DataFfi = mv8.data_node_from_js(data);
 
-        let n_children = child_data.len();
-        let child_data: Vec<ArrayData> = match dt.clone() {
-            DataType::List(field) => {
-                let child: mv8::Value<'_> = child_data.get(0)?;
-                Ok(vec![self.array_data_from_js(
+        // This target length is used because the JS repr does not mirror buffer building as Rust
+        // Arrow and pyarrow.
+        let target_len = len.unwrap_or(data.len);
+
+        // TODO: Extra copies when generating buffers, because JS Arrow doesn't align things
+        //   properly. (Due to which buffer capacities are currently unused.)
+        let mut builder = ArrayData::builder(field.data_type().clone());
+
+        match field.data_type() {
+            DataType::Boolean => unsafe {
+                let mut boolean_builder = BooleanBufferBuilder::new(target_len);
+                let values = slice::from_raw_parts(data.buffer_ptrs[0], target_len);
+                boolean_builder.append_packed_range(0..target_len, values);
+
+                builder = builder.add_buffer(boolean_builder.finish());
+            },
+            DataType::UInt16 => unsafe {
+                let values = slice::from_raw_parts(data.buffer_ptrs[0] as *const u16, target_len);
+                builder = builder.add_buffer(Buffer::from_slice_ref(&values));
+            },
+            DataType::UInt32 => unsafe {
+                let values = slice::from_raw_parts(data.buffer_ptrs[0] as *const u32, target_len);
+                builder = builder.add_buffer(Buffer::from_slice_ref(&values));
+            },
+            DataType::Float64 => unsafe {
+                let values = slice::from_raw_parts(data.buffer_ptrs[0] as *const f64, target_len);
+                builder = builder.add_buffer(Buffer::from_slice_ref(&values));
+            },
+            DataType::Utf8 => unsafe {
+                // TODO: Use `data.len` or target_len?
+                //       (In practice, target_len has worked for a long time,
+                //       though that's not an ideal reason to use it. Maybe
+                //       `data.len` would also work.)
+                let offsets =
+                    slice::from_raw_parts(data.buffer_ptrs[0] as *const i32, target_len + 1);
```

Comment:
The lack of explanation in this file is worrying to me. Especially the inconsistency with `target_len` and `target_len + 1` and also the TODO above.

I appreciate you migrated this code and did not write most of it, so perhaps it's unrealistic to expect you to be able to do so, but can you figure it out and document it better? 

If it's a lot of work then lets add a TODO to improve the readability and explanation in this code. I specifically want to know at a glace why we're using target_len instead of data.len (e.g. testing the TODO above which we might need to wait until tests pass and a merge with main), and also why we're adding +1 to the len sometimes, etc.